common/grammar-parser.cpp

```diff
@@ -137,6 +152,87 @@ namespace grammar_parser {
             bool                                 is_nested) {
         size_t last_sym_start = out_elements.size();
         const char * pos = src;
+
+        auto handle_repetitions = [&](int min_times, int max_times) {
+
+            if (last_sym_start == out_elements.size()) {
+                throw std::runtime_error(std::string("expecting preceding item to */+/?/{ at ") + pos);
+            }
+
+            // apply transformation to previous symbol (last_sym_start to end) according to
+            // the following rewrite rules:
+            // S*     --> S{0,}
+            // S+     --> S{1,}
+            // S?     --> S{0,1}
+            // S{m,n} --> S'     ::= Scopy Scopy Scopy... (m times) S(n-m)
+            //            Scopy  ::= S
+            //            S(x)   ::= Scopy S(x-1) |
+            //            S(x-1) ::= Scopy S(x-2) |
+            //            S(1)   ::= Scopy |
```

Comment:
```suggestion
            //            S(x-1) ::= Scopy S(x-2) |
            //            S(x-2) ::= ... (n-m times, down to S(1))
            //            S(1)   ::= Scopy |
```

Reply:
This part of the docs was really confusing me until I read the code until I realized that this loops and continues. Might be helpful to make a note about this.

Reply:
Made this a bit more verbose:

```c++
            // S{m,n} --> S S S (m times) S'(n-m)
            //            S'(x)   ::= S S'(x-1) |
            //            (... x definitions of these S' rules ...)
            //            S'(1)   ::= S |
```